lis2dw12: add BW filter, low noise, threshold HP_REF and autoreset

maximevince opened the pull request:
Added support for a few of the more advanced features present in the LIS2DW12 accelero:
* Digital filtering bandwidth selection
* Wakeup threshold trigger
* High-Pass reference mode
* High-Pass reference auto-reset on wakeup interrupt

Let me know if the split between DeviceTree config (like the bw_filt value) and the Kconfig defines is okay.

mbolivar-nordic commented on drivers/sensor/lis2dw12/Kconfig:
```diff
@@ -59,4 +59,30 @@ config LIS2DW12_TAP
+config LIS2DW12_BW_FILT
+	bool "Digital filtering bandwidth selection"
+	help
+	  Specify BW_FILT value to be used in the DeviceTree.
```
What's going on here? This is in Kconfig, not DT.

avisconti replied:
Yes, everything should be done in DT. I don't see changes in the dts/binding. Am I missing something?

avisconti commented on drivers/sensor/lis2dw12/lis2dw12.c:
```diff
@@ -289,13 +379,26 @@ static int lis2dw12_init(const struct device *dev)
 
+#ifdef CONFIG_LIS2DW12_BW_FILT
+    LOG_INF("Setting filter bandwidth: %d", (int)cfg->bw_filt);
+    lis2dw12_filter_bandwidth_set(ctx, cfg->bw_filt);
```
Wherte is bw_filt assigned? Should be taken from DT right?

maximevince replied:
This part was missing, indeed. I have pushed a new commit where this is fixed.

avisconti replied:
OK, and I think you can get rid of CONFIG_LIS2DW12_BW_FILT in Kconfig.
Again, the default will be 0 if not defined in a particular driver instance.

avisconti commented:
@maximevince 
Please also change the commit header format

maximevince commented:
Thanks for the feedback, I will make sure to process the changes requested. It will take some time, as I am on leave now.

maximevince commented:
> @maximevince Please also change the commit header format

Excuse my ignorance, but could you give me an example of a correct commit header format?

avisconti commented:
> > @maximevince Please also change the commit header format
> 
> Excuse my ignorance, but could you give me an example of a correct commit header format?
> 
> EDIT: I have guessed it was the `drivers/sensor: lis2dw12:` prefix you were referring to, so I have added those. Let me know if it's okay now.

Please take a look at https://docs.zephyrproject.org/latest/contribute/index.html#commit-guidelines
Basically, you should have:

```
commit title
<blank line>
commit body (multiple lines)
<blank line>
signature
```

avisconti commented on drivers/sensor/lis2dw12/lis2dw12.c:
```diff
@@ -289,13 +379,26 @@ static int lis2dw12_init(const struct device *dev)
 	}
 #endif /* CONFIG_LIS2DW12_TRIGGER */
 
+#ifdef CONFIG_LIS2DW12_THRESHOLD_HP_REF
```
Again, define it in DT.

avisconti reviewed (commented):
I think you should also re-organize the commits per feature, like wake-up support, tap support, ...

maximevince commented on drivers/sensor/lis2dw12/lis2dw12.c:
```diff
@@ -275,6 +359,12 @@ static int lis2dw12_init(const struct device *dev)
 		return -EIO;
 	}
 
+#ifdef CONFIG_LIS2DW12_LOW_NOISE
```
That makes sense!

maximevince committed:
drivers/sensor: lis2dw12: add bw_filt support

Add bandwidth filter support. (BW_FILT in CTRL6)
Value is configurable through DT per instance.

Signed-off-by: Maxime Vincent <maxime@veemax.be>

maximevince committed:
drivers/sensor: lis2dw12: add low_noise support

Add low_noise support. (LOW_NOISE in CTRL6)
Value is configurable through DT per instance.

Signed-off-by: Maxime Vincent <maxime@veemax.be>

maximevince committed:
drivers/sensor: lis2dw12: add fds + hp_ref support

Add FDS (Filtered Data Type Selection) + High-Pass reference mode support
(FDS in CTRL6, HP_REF_MODE in CTRL7)
Values are configurable through DT per instance.

Signed-off-by: Maxime Vincent <maxime@veemax.be>

maximevince committed:
drivers/sensor: lis2dw12: add threshold interrupt support

Add optional threshold interrupt support.
Implemented using SENSOR_TRIG_THRESHOLD sensor trigger type.
The features can be optionally enabled through Kconfig,
or disabled for smaller code size.

Signed-off-by: Maxime Vincent <maxime@veemax.be>

maximevince committed:
drivers/sensor: lis2dw12: add drdy pulsed/latched config

Add DT option to configure the data ready interrupt mode.
Latched is the default; pulsed can be enabled through
the drdy-pulsed DT, if desired.

Signed-off-by: Maxime Vincent <maxime@veemax.be>

maximevince commented:
@avisconti : I have restructured the commits to contain one (set of) feature(s) per commit, and re-did the commits messages. Commits are now signed and GPG-signed as well.

I have left one Kconfig option in there: `CONFIG_LIS2DW12_THRESHOLD`, which allows to disable this portion of the code, in case a smaller code size is desired and the threshold feature is not used.

All other config options have been moved to DT.

mbolivar-nordic commented on drivers/sensor/lis2dw12/lis2dw12.c:
```diff
@@ -289,6 +289,9 @@ static int lis2dw12_init(const struct device *dev)
 		return ret;
 	}
 
+	LOG_INF("bandwidth filter is %d", (int)cfg->bw_filt);
```
LOG_INF seems a bit much; why did you select this instead of LOG_DBG like the other calls in this same function?

mbolivar-nordic commented on drivers/sensor/lis2dw12/lis2dw12.c:
```diff
@@ -319,6 +319,20 @@ static int lis2dw12_init(const struct device *dev)
 	}
 #endif /* CONFIG_LIS2DW12_TRIGGER */
 
+	LOG_INF("high pass reference mode is %d", (int)cfg->hp_ref_mode);
```
Similar comments in this patch -- LOG_INF seems a bit noisy for these, no?

maximevince replied:
Changed them to LOG_DBG. Totally makes sense!

mbolivar-nordic commented on drivers/sensor/lis2dw12/lis2dw12.h:
```diff
@@ -70,6 +70,8 @@ struct lis2dw12_device_config {
 	uint8_t range;
 	uint8_t bw_filt;
 	bool low_noise;
+	bool hp_filter_path;
```
It's @avisconti 's driver so this is not a blocking request, but it seems like you could save a little flash by bit-packing these.

mbolivar-nordic commented on drivers/sensor/lis2dw12/lis2dw12_trigger.c:
```diff
@@ -67,6 +67,19 @@ static int lis2dw12_enable_int(const struct device *dev,
+	 * attributes.
+	 */
+	case SENSOR_TRIG_THRESHOLD:
+		LOG_INF("Setting int1_wu: %d\n", enable);
```
You get the idea about LOG_INF being probably too noisy -- I'll stop mentioning it

mbolivar-nordic reviewed (dismissed):
Devicetree bindings changes are well done. Nice job.

A few random comments but none are blocking. OK for me on the DT side, so I'm approving under the assumption that @avisconti should +1 before merge

avisconti commented:
Very good job indeed!

maximevince commented:
Okay, one more push where I changed all of the added `LOG_INF`s to `LOG_DBG`s.
If it's okay for you guys, it can be merged!

MaureenHelm commented:
> Okay, one more push where I changed all of the added `LOG_INF`s to `LOG_DBG`s.
> If it's okay for you guys, it can be merged!

These changes were made in the wrong commit.

There's also a compliance check that needs to be fixed. Everything else looks great

maximevince commented:
> > Okay, one more push where I changed all of the added `LOG_INF`s to `LOG_DBG`s.
> > If it's okay for you guys, it can be merged!
> 
> These changes were made in the wrong commit.
> 
> There's also a compliance check that needs to be fixed. Everything else looks great

Thanks @MaureenHelm . Issues have been addressed!

gmarull commented on drivers/sensor/lis2dw12/lis2dw12.c:
```diff
@@ -289,6 +289,9 @@ static int lis2dw12_init(const struct device *dev)
 		return ret;
 	}
 
+	LOG_DBG("bandwidth filter is %d", (int)cfg->bw_filt);
```
just use `%u`, or, `PRIu8`

gmarull commented on drivers/sensor/lis2dw12/lis2dw12.c:
```diff
@@ -236,6 +236,20 @@ static int lis2dw12_set_power_mode(const struct device *dev,
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	lis2dw12_ctrl6_t ctrl6;
+
+	if (lis2dw12_read_reg(ctx, LIS2DW12_CTRL6, (uint8_t *)&ctrl6, 1)) {
+		return -EIO;
+	}
```
please propagate error codes if you're not handling them,

```
ret = f();
if (ret < 0) {
    return ret;
}
```

applies to others

maximevince replied:
I just mirrored what the rest of the driver was doing.
I'll refactor my part.

gmarull commented on drivers/sensor/lis2dw12/lis2dw12.c:
```diff
@@ -157,11 +158,81 @@ static int lis2dw12_config(const struct device *dev, enum sensor_channel chan,
+
+	uint32_t thr_mg = abs(sensor_ms2_to_mg(val));
+
+	/* Check maximum value: depends on current FS value */
+	if (thr_mg >= (2U << range) * 1000) {
+		return -EINVAL;
+	}
+
+	/* The threshold is applied to both positive and negative data:
+	 * for a wake-up interrupt generation at least one of the three axes must be
+	 * bigger than the threshold. */
+	int lsb_mg = ((2U << range) * 1000) / 64;; /* range = 2/4/8/16G */
+
+	reg = (thr_mg + (lsb_mg / 2)) / lsb_mg;
```
magic constants here, please, define or document them.

gmarull commented on drivers/sensor/lis2dw12/Kconfig:
```diff
@@ -59,4 +59,8 @@ config LIS2DW12_TAP
 endif # LIS2DW12_TRIGGER
 
+config LIS2DW12_THRESHOLD
+	bool "Wakeup threshold trigger (via interrupt)"
```
please add some help here

maximevince replied:
Ok, done!

gmarull commented on drivers/sensor/lis2dw12/lis2dw12.c:
```diff
@@ -157,11 +158,81 @@ static int lis2dw12_config(const struct device *dev, enum sensor_channel chan,
 {
+#if CONFIG_LIS2DW12_THRESHOLD
+	switch (attr) {
+	case SENSOR_ATTR_UPPER_THRESH:
```
missing `__fallthrough`

maximevince replied:
`__fallthrough` for consecutive cases in a switch? (without any other lines of code in between)
Sure, I can do that, but:
- doesn't that hurt readability more, than help it?
- I can't see that being done almost anywhere else in `drivers/sensor`

gmarull reviewed (dismissed):
some nitpicks

maximevince commented on drivers/sensor/lis2dw12/lis2dw12.c:
```diff
@@ -157,11 +158,81 @@ static int lis2dw12_config(const struct device *dev, enum sensor_channel chan,
 }
 
+
+static inline int32_t sensor_ms2_to_mg(const struct sensor_value *ms2)
```
Not sure I understand the suggestion.
Using `sensor_ms2_to_g` would make me loose a lot of precision; everything after the decimal dot, right?
Or are you suggesting to scale the input to `sensor_ms2_to_g` before calling the conversion?

maximevince commented:
Addressed a few style issues + fixed a bug in setting the threshold interrupt and value.

maximevince commented:
Anything missing from my side to get this merged? @avisconti @mbolivar-nordic @gmarull @MaureenHelm

mbolivar-nordic commented:
@gmarull is out on leave, so I will dismiss his review if @MaureenHelm or a sensor collaborator can give this another +1

avisconti commented:
> @avisconti : I have restructured the commits to contain one (set of) feature(s) per commit, and re-did the commit messages. Commits are now signed and GPG-signed as well.
> 
> I have left one Kconfig option in there: `CONFIG_LIS2DW12_THRESHOLD`, which allows to disable this portion of the code when not selected, which results in smaller code size, when the threshold feature is not used.
> 
> All other config options have been moved to DT.

But once enabled it is enabled for all driver instance. I usually have same comment here, which is that there might be the need to have one instance configured in one way and the other not. Isn't it?
Moreover the LIS2DW12 is a plain accelerometer, which means that there may be lot of application using multiple instances.

avisconti commented on drivers/sensor/lis2dw12/Kconfig:
```diff
@@ -59,4 +59,16 @@ config LIS2DW12_TAP
+	  be bigger than the threshold. See ST AN5038 for more details about
+	  the feature.
+
+
```
Please reconsider to have also this one in DT, or convince me that is not needed.

maximevince replied:
Hi @avisconti .
The Kconfig choice here is about the firmware feature: code compiled in or not,
not about enabling/configuring this feature for a specific LIS2DW12 instance.

My board also has two LIS2DW12's on it, so I can still choose to only configure one in threshold mode. The user still needs to call `lis2dw12_attr_set_thresh()` on their specific instance to configure the feature.

It was mainly done from a code-size perspective, hope this is acceptable.

avisconti replied:
> Hi @avisconti . The Kconfig choice here is about the firmware feature: code compiled in or not, not about enabling/configuring this feature for a specific LIS2DW12 instance.
> 
> My board also has two LIS2DW12's on it, so I can still choose to only configure one in threshold mode. The user still needs to call `lis2dw12_attr_set_thresh()` on their specific instance to configure the feature.
> 
> It was mainly done from a code-size perspective, hope this is acceptable.

Oh yes, sorry. It is called at runtime.
My bad...

avisconti reviewed (changes requested):
Everything is good but the THRESHOLD in Kconfig.
Nevertheless, I'm ready to change my mind if this choice is supported by good arguments. In my opinion the flexibility to select different configs for different driver instances is more important than having a smaller code size.

MaureenHelm reviewed (approved):
@gmarull please revisit
